community/kernel/src/main/java/org/neo4j/kernel/impl/api/StateHandlingStatementOperations.java

```diff
@@ -397,37 +399,69 @@ public RelationshipIterator relationshipsGetAll( KernelStatement state )
     @Override
     public boolean nodeAddLabel( KernelStatement state, long nodeId, int labelId ) throws EntityNotFoundException
     {
-        if ( nodeHasLabel( state, nodeId, labelId ) )
+        try ( NodeCursor node = nodeCursor( state, nodeId ) )
         {
-            // Label is already in state or in store, no-op
-            return false;
-        }
+            if ( !node.next() )
+            {
+                throw new EntityNotFoundException( EntityType.NODE, nodeId );
+            }
 
-        state.txState().nodeDoAddLabel( labelId, nodeId );
-        for ( Iterator<DefinedProperty> properties = nodeGetAllProperties( state, nodeId ); properties.hasNext(); )
-        {
-            DefinedProperty property = properties.next();
-            indexUpdateProperty( state, nodeId, labelId, property.propertyKeyId(), null, property );
+            try ( LabelCursor labels = node.labels() )
+            {
+                if ( labels.seek( labelId ) )
+                {
+                    // Label is already in state or in store, no-op
+                    return false;
+                }
+            }
+
+            state.txState().nodeDoAddLabel( labelId, nodeId );
+
+            try ( PropertyCursor properties = node.properties() )
+            {
+                while ( properties.next() )
+                {
+                    DefinedProperty property = Property.property( properties.propertyKeyId(), properties.value() );
+                    indexUpdateProperty( state, nodeId, labelId, property.propertyKeyId(), null, property );
```

Comment:
Yeah, that code needs to be updated, but is out of scope for this PR. There's another card for it.